src/models/order_rewards_schema.py

```diff
@@ -0,0 +1,69 @@
+"""Model for Order Rewards Data"""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any
+
+from pandas import DataFrame
+
+
+@dataclass
+class OrderRewards:
+    """
+    This class provides a transformation interface for the Dataframe we fetch from the orderbook
+    Within are several methods for converting from the Dataframe to list[DuneRecords]
+    """
+
+    order_uid: str
+    tx_hash: str
+    solver: str
+    data: dict[str, Any]
+
+    @classmethod
+    def from_pdf(cls, rewards_df: DataFrame) -> list[OrderRewards]:
+        """Multi Record Constructor from Pandas Dataframe"""
+        return [
+            cls(
+                order_uid=row["order_uid"],
+                tx_hash=row["tx_hash"],
+                solver=row["solver"],
+                data={
+                    # This seems like a hack. I need to experiment with Decimal type
+                    # CAST("18446744073709001111111111" AS DECIMAL(38,0));
+                    # but this can be done on the spellbook side.
+                    "surplus_fee": str(row["surplus_fee"]),
+                    "amount": float(row["amount"]),
+                    "safe_liquidity": row["safe_liquidity"],
+                },
+            )
+            for row in rewards_df.to_dict(orient="records")
+        ]
+
+    def as_dune_record(self) -> dict[str, Any]:
+        """Convert to Dune Record"""
+        return {
+            "order_uid": self.order_uid,
+            "tx_hash": self.tx_hash,
+            "solver": self.solver,
+            "data": self.data,
+        }
+
+    @classmethod
+    def from_pdf_to_dune_records(cls, rewards_df: DataFrame) -> list[dict[str, Any]]:
+        """
+        The above two methods require two iterations of the entire record set
+        while this one combines both and requires only one.
+        """
+        return [
+            {
+                "order_uid": row["order_uid"],
+                "tx_hash": row["tx_hash"],
+                "solver": row["solver"],
+                "data": {
+                    "surplus_fee": str(row["surplus_fee"]),
+                    "amount": float(row["amount"]),
+                    "safe_liquidity": row["safe_liquidity"],
+                },
+            }
+            for row in rewards_df.to_dict(orient="records")
```

Comment:
I find all this dataframe manipulation very clunky and I would be open to alternative suggestions here.

The two methods above, for example are not even used in this project because I think two iterations over the same record set is silly and unnecessary. This data class only exists to transform the dataframe into the type we want to stream.

Reply:
Update: I have remove all the unused code: It is here now in case anyone cares

```py

    @classmethod
    def from_pdf(cls, rewards_df: DataFrame) -> list[OrderRewards]:
        """Multi Record Constructor from Pandas Dataframe"""
        return [
            cls(
                order_uid=row["order_uid"],
                tx_hash=row["tx_hash"],
                solver=row["solver"],
                data={
                    "surplus_fee": str(row["surplus_fee"]),
                    "amount": float(row["amount"]),
                    "safe_liquidity": row["safe_liquidity"],
                },
            )
            for row in rewards_df.to_dict(orient="records")
        ]

    def as_dune_record(self) -> dict[str, Any]:
        """Convert to Dune Record"""
        return {
            "order_uid": self.order_uid,
            "tx_hash": self.tx_hash,
            "solver": self.solver,
            "data": self.data,
        }

```

Reply:
The only update I could think of about this is that you could vectorize it like this:

```python
        rewards_df["data"] = rewards_df.apply(
            lambda row:
                {
                    "surplus_fee": str(row["surplus_fee"]),
                    "amount": float(row["amount"]),
                    "safe_liquidity": row["safe_liquidity"],
                }, axis=1
        )
        return rewards_df[["order_uid", "tx_hash", "solver", "data"]].to_dict(
            orient="records"
        )
```

Let me know what you think about it!

Reply:
So this is definitely the kind of suggestion I was hoping for (being able to create the "data" column in pandas syntax) however, it looks like this also requires two iterations over the complete data frame (one to build the new column and then one to export the thing in the format). For efficiency reasons, I think I will stick with the current approach. Thanks for this!